File: Lab3/Solution/fileReverse.c
//-----------------------------------------------------------------------------
// FileReverse.c
// Reverses all tokens in a file and outputs them to another file
// Solution to Lab3 in CMPS 12B/M 
// Code reuse from other lab 3 documents. 
// 
// Usage: this should be compiled into executable FileReverse.
// The command "./FileReverse <in> <out>" will produce a file <out> that, in a separate line, has the reverse of each word of <in>
//
// Matt Bryson, Jan 2016. C. Seshadhri, Sept 2017.
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>

//Function definitions 
void stringReverse(char* s);

int main(int argc, char* argv[])
{
	FILE* in;  /* file handle for input */  
	FILE* out; /* file handle for output */
	char word[256]; /* char array to store words from input file */
	
    if(argc < 3){
       printf("Usage: %s <input file> <output file>\n", argv[0]);
       exit(EXIT_FAILURE);
    }
	  
	/* open input file for reading */
	in = fopen(argv[1], "r");
	if( in==NULL )
	{
		printf("Unable to read from file %s\n", argv[1]);
		exit(EXIT_FAILURE);
	}

	/* open output file for writing */
	out = fopen(argv[2], "w");
	if( out==NULL )
	{
		printf("Unable to write to file %s\n", argv[2]);
		exit(EXIT_FAILURE);
	}
	
	/* read words from input file, print on separate lines to output file*/
	while( fscanf(in, " %s", word) != EOF )
	{
		//printf("%s\n", word);
		stringReverse(word);
		fprintf(out, "%s\n", word);
	}

	/* close input and output files */
	fclose(in);
	fclose(out);
	
	return(EXIT_SUCCESS);
	
}

// function to reverse string. 
// Input: string s
// Output: none. The string s is modified into its reverse.

void stringReverse(char* s)
{
	int length, halfway;
	if(s != NULL) // if string is null, do nothing
	{
		length = strlen(s); 
	}
	else
	{
		return;
	}
	
    halfway = (length-1)/2; // index at half the length of s
   
    for(int i = 0; i <= halfway; i++) //loop up to the halfway point
	{
        int to_swap = length-1-i; //index where ith char should end up
		char temp = s[i];  // swap s[i] and s[to_swap]
		s[i] = s[to_swap];
		s[to_swap] = temp;
	}
}

File: Lab2/Solution/FileReverse.java
//-----------------------------------------------------------------------------
// FileReverse.java
// Reverses all tokens in a file and outputs them to another file
// Solution to Lab2 in CMPS 12B/M 
// Code reuse from other lab 2 documents. 
//
//
// Usage: ideally, this should be compiled into a jar file. 
// The command "java -jar FileReverse.jar <in> <out>" will produce a file <out> that, in a separate line, has the reverse of each word of <in>
//
// Matt Bryson, Oct 2016. C. Seshadhri, Sept 2017.
//-----------------------------------------------------------------------------
import java.io.*;
import java.util.Scanner;

class FileReverse{
   public static void main(String[] args) throws IOException{

      // check number of command line arguments is at least 2
      if(args.length < 2){
         System.out.println("Usage: java -jar FileReverse.jar <input file> <output file>");
         System.exit(1);
      }

      // open files
      Scanner in = new Scanner(new File(args[0]));
      PrintWriter out = new PrintWriter(new FileWriter(args[1]));

      // read lines from in, extract and print tokens from each line
      while( in.hasNextLine() ){
         // trim leading and trailing spaces, then add one trailing space so 
         // split works on blank lines
         String line = in.nextLine().trim() + " "; 

         // split line around white space 
         String[] token = line.split("\\s+");  

         // print out tokens       
         int n = token.length;
         for(int i=0; i<n; i++){
            out.println(stringReverse(token[i]));
         }
      }

      // close files
      in.close();
      out.close();
   }
   
// function to reverse a string
   public static String stringReverse(String s)
   {
	   int length = s.length(); 
	   char [] reversed = new char [length]; // set up output array of correct length
	   
	   int y = length - 1;  // initialize to last position in s
	   for(int i = 0; i < length; i++) // loop to fill in reversed
	   {
		   reversed [i] = s.charAt(y); // copy into reversed, and decrement y. Basically, loop is going backwards in s
		   y--; 
	   }
	   
	   return String.valueOf(reversed);
   }
   
}
